image_proc/src/crop_non_zero.cpp

```diff
@@ -0,0 +1,149 @@
+*     copyright notice, this list of conditions and the following
+*     disclaimer in the documentation and/or other materials provided
+*     with the distribution.
+*   * Neither the name of the Willow Garage nor the names of its
+*     contributors may be used to endorse or promote products derived
+*     from this software without specific prior written permission.
+* 
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+#include <rclcpp/rclcpp.hpp>
+#include <image_transport/image_transport.h>
+#include <boost/thread.hpp>
+#include <cv_bridge/cv_bridge.h>
+#include "crop_non_zero.hpp"
+//#include <algorithm> // for std::max_element
+
+namespace image_proc
+{
+
+namespace enc = sensor_msgs::image_encodings;
+
+CropNonZeroNode::CropNonZeroNode(const rclcpp::NodeOptions& options) : Node("CropNonZeroNode", options)
+{
+  auto parameter_change_cb =
+    [this](std::vector<rclcpp::Parameter> parameters) -> rcl_interfaces::msg::SetParametersResult {
+    auto result = rcl_interfaces::msg::SetParametersResult();
+    result.successful = true;
+    for (auto parameter : parameters)
+    {
+      if (parameter.get_name() == "camera_namespace")
+      {
+        camera_namespace_ = parameter.as_string();
+        RCLCPP_INFO(get_logger(), "camera_namespace: %s ", camera_namespace_.c_str());
+        break;
+      }
+    }
+    image_pub_ = camera_namespace_ + "/image";
+    connectCb();
+    // Make sure we don't enter connectCb() between advertising and assigning to pub_depth_
+    std::lock_guard<std::mutex> lock(connect_mutex_);
+    RCLCPP_INFO(this->get_logger(), "publish: %s", image_pub_.c_str());
+    pub_ = image_transport::create_publisher(this, image_pub_);
+    return result;
+  };
+  this->declare_parameter("camera_namespace");
+
+  rclcpp::Parameter parameter;
+  if (rclcpp::PARAMETER_NOT_SET != this->get_parameter("camera_namespace", parameter)) {
+    parameter_change_cb(this->get_parameters({"camera_namespace"}));
+  }
+
+  this->set_on_parameters_set_callback(parameter_change_cb);
+}
+
+// Handles (un)subscribing when clients (un)subscribe
+void CropNonZeroNode::connectCb()
```

Comment:
So connectCb is only called in this lambda, and the next line in the lambda relocks the same mutex. It would be better to either:

a) put the connectCb stuff  inside of the lambda, if there's nothing being otherwise exposed from the outside that needs it
b) only have the mutex in the lambda before the connectCb is called so we're not having 2 locks toggling quickly

Reply:
Went for b) (thinking there was some logic for remove a subscription when the subscriber number were 0 that might be added in the future)